pandas/core/arrays/datetimelike.py

```diff
@@ -1303,16 +1303,18 @@ def _sub_periodlike(self, other: Period | PeriodArray) -> npt.NDArray[np.object_
         self._check_compatible_with(other)
 
         other_i8, o_mask = self._get_i8_values_and_mask(other)
-        new_i8_data = add_overflowsafe(self.asi8, np.asarray(-other_i8, dtype="i8"))
-        new_data = np.array([self.freq.base * x for x in new_i8_data])
-
+        new_data = add_overflowsafe(
+            self.asi8, np.asarray(-other_i8, dtype="i8")
+        ).astype(object)
```

Comment:
Not a great solution if this always needs to be cast to `object`

Reply:
Why not? The comment says "we return an object-dtype ndarray" so it makes sense to cast it to `object`.
I thought this is better than creating a new array which will be object dtype (old behavior)

Reply:
I agree this seems suboptimal, and that comment seems pretty far detached from where the object conversion happens right? I agree this is a bit suspect; the old code seems to use primitive types so overall this is adding a good deal of overhead.

Do you know where the cast to object is actually occurring?

Reply:
```python
new_data = np.array([self.freq.base * x for x in new_i8_data])
```
Here, `self.freq.base` is what causes `new_data` to be object dtype

Reply:
Running a quick check, it seems my solution has the same performance as the old code.
```python
In [1]: import pandas as pd
In [2]: pr = pd.period_range(start='2017-01-01', periods=100, freq="D")
In [3]: %timeit pd.Series(pr).diff()
280 µs ± 1.14 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```
Compared to old code:
```python
In [3]: %timeit pd.Series(pr).diff()
\pandas\core\arrays\datetimelike.py:1307: RuntimeWarning: overflow encountered in scalar multiply
  new_data = np.array([self.freq.base * x for x in new_i8_data])
357 µs ± 1.85 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```